### PR TITLE
Add negative log likelihood

### PR DESCRIPTION
**Pull request by Descartess**
Reference #127

### CODE REVIEW

**Comment by Descartess** [comment]
I signed it

File: Tests/TensorFlowTests/LossTests.swift [hunk 1]
```diff
@@ -67,6 +67,18 @@ final class LossTests: XCTestCase {
 
         assertElementsEqual(expected: expectedGradients, actual: gradients)
     }
+    
+    func testNegativeLogLikelihood() {
+        let predicted = Tensor<Float>(shape: [2, 4], scalars: [1, 2, 3, 4, 5, 6, 7, 8])
```
Comment (Shashi456) [inline Tests/TensorFlowTests/LossTests.swift]: Can you also have an expected variable which just checks if the outputs are right? You can add the Assert equal to the same test case.

File: Tests/TensorFlowTests/LossTests.swift [hunk 2]
```diff
@@ -190,6 +202,7 @@ final class LossTests: XCTestCase {
         ("testMeanSquaredErrorGrad", testMeanSquaredErrorGrad),
         ("testMeanSquaredLogarithmicError", testMeanSquaredLogarithmicError),
         ("testMeanAbsoluteError", testMeanAbsoluteError),
+        ("testNegativeLogLikelihood",testNegativeLogLikelihood),
```
Comment (Shashi456) [inline Tests/TensorFlowTests/LossTests.swift]:
```suggestion
        ("testNegativeLogLikelihood", testNegativeLogLikelihood),
```

File: Tests/TensorFlowTests/LossTests.swift [hunk 3]
```diff
@@ -67,6 +67,18 @@ final class LossTests: XCTestCase {
...
+    
+    func testNegativeLogLikelihood() {
+        let predicted = Tensor<Float>(shape: [2, 4], scalars: [1, 2, 3, 4, 5, 6, 7, 8])
+        let expectedGradients = Tensor<Float>(
+            shape: [2, 4],
+            scalars: [-0.125, -0.0625, -0.0416666, -0.03125, -0.025, -0.0208333, -0.017857, -0.015625])
+        let gradients = gradient(
+            at: predicted,
+            in: { negativeLogLikelihood(predicted: $0) })
+
```
Comment (Shashi456) [inline Tests/TensorFlowTests/LossTests.swift]:
```suggestion
```
Remove the extra line.

File: Tests/TensorFlowTests/LossTests.swift [hunk 4]
```diff
@@ -67,6 +67,18 @@ final class LossTests: XCTestCase {
...
+    func testNegativeLogLikelihood() {
+        let predicted = Tensor<Float>(shape: [2, 4], scalars: [1, 2, 3, 4, 5, 6, 7, 8])
+        let expectedGradients = Tensor<Float>(
+            shape: [2, 4],
+            scalars: [-0.125, -0.0625, -0.0416666, -0.03125, -0.025, -0.0208333, -0.017857, -0.015625])
+        let gradients = gradient(
+            at: predicted,
+            in: { negativeLogLikelihood(predicted: $0) })
+        
+        assertElementsEqual(expected: expectedGradients, actual: gradients)
```
Comment (Descartess) [inline Tests/TensorFlowTests/LossTests.swift]: > Can you also have an expected variable which just checks if the outputs are right? You can add the Assert equal to the same test case.

@Shashi456 Correct me if am wrong, but I believe that is being done here.

Comment (Shashi456) [reply Tests/TensorFlowTests/LossTests.swift]: You are actually checking the gradients here. You could name the loss function `testNegativeLogLikelihoodGrad`.

**Comment by Shashi456** [comment]
@Descartess good job, LGTM. I'm not very sure about the naming though, a variety of frameworks name this NLLloss as well. perhaps @rxwei or @eaplatanios might be of help here.

File: Sources/TensorFlow/Loss.swift [hunk 5]
```diff
@@ -53,6 +53,17 @@ public func meanAbsoluteError<Scalar: TensorFlowFloatingPoint>(
...
 }
 
+/// Returns the negative logarithmic likelihood of the predictions.
+///
+/// - Parameters:
+///   - predicted: Predicted outputs from a neural network.
+@differentiable(wrt: predicted)
+public func negativeLogLikelihood<Scalar: TensorFlowFloatingPoint>(
+    predicted: Tensor<Scalar>
+    ) -> Tensor<Scalar> {
```
Comment (eaplatanios) [inline Sources/TensorFlow/Loss.swift]:
```suggestion
) -> Tensor<Scalar> {
```

**Comment by rxwei** [comment]
Thanks for putting this together @Descartess.  I agree with @eaplatanios and I think functions like this could complicate our API surface in that it's just a negative log and does not have the same semantics as prior art like [NLLLoss](https://pytorch.org/docs/stable/nn.html#nllloss). So I'd suggest dropping this function for now.

**Comment by Descartess** [comment]
👍
